packages/permission-controller/src/Caveat.ts

```diff
@@ -105,6 +106,62 @@ export type CaveatValidator<ParentCaveat extends CaveatConstraint> = (
   target?: string,
 ) => void;
 
+/**
+ * A function that merges two caveats of the same type. The values
+ * must be merged in the fashion of a right-biased union.
+ *
+ * @see `ARCHITECTURE.md` for more details.
+ * @template ParentCaveat - The caveat type associated with this merger.
+ * @param leftCaveat - The left-hand caveat.
+ * @param rightCaveat - The right-hand caveat.
+ * @returns The merged caveat.
+ */
+export type CaveatMerger<ParentCaveat extends CaveatConstraint> = (
+  leftCaveat: ParentCaveat,
+  rightCaveat: ParentCaveat,
+) => ParentCaveat;
+
+/**
+ * A function that merges two caveat values of the same type. The values
+ * must be merged in the fashion of a right-biased union.
+ *
+ * @see `ARCHITECTURE.md` for more details.
+ * @template Value - The type of the values to merge.
+ * @param leftValue - The left-hand value.
+ * @param rightValue - The right-hand value.
+ * @returns The merged value.
+ */
+export type CaveatValueMerger<Value extends Json> = (
+  leftValue: Value,
+  rightValue: Value,
+) => Value;
+
+/**
+ * Makes a {@link CaveatMerger} function for a specific caveat type.
+ *
+ * @see `ARCHITECTURE.md` for more details.
+ * @param mergeValues - The function to merge the values of two caveats.
+ * @returns The {@link CaveatMerger} function.
+ */
+export function makeCaveatMerger<ParentCaveat extends CaveatConstraint>(
+  mergeValues: CaveatValueMerger<ParentCaveat['value']>,
```

Comment:
```suggestion
  mergeValuesFn: CaveatValueMerger<ParentCaveat['value']>,
```
Nit, seems a bit more explicit, I think.

Reply:
Perhaps, however our naming convention for functions states that they should read like a verb, and appending `Fn` turns it into a noun, which is problematic when we call `mergeValues` in the body of the function.

Reply:
Yeah but then in this case `mergeValues` also sounds like a noun. perhaps `mergeCaveatValues`?

Reply:
I think it's more of an english language thing haha, `mergeValues` can be understood to also mean values that need to be merged 😄

Reply:
Haha, you are absolutely right, but in the context as a parameter of type `CaveatValueMerger` in a function named `makeCaveatMerger`, the only reasonable interpretation of the name is that it's a function, _and_ although English permits either interpretation, I claim that `mergeValues` is _only_ a valid name for a function under our naming conventions.

Reply:
haha ok fair point :)